core/frontend/src/common/render/GraphicAssembler.ts

```diff
@@ -0,0 +1,462 @@
+    if (2 === points.length && points[0].isAlmostEqual(points[1]))
+      this[_accumulator].addPointString(points, this.getLinearDisplayParams(), this.placement);
+    else
+      this[_accumulator].addLineString(points, this.getLinearDisplayParams(), this.placement);
+  }
+
+  /**
+   * Appends a 2d line string to the builder.
+   * @param points Array of vertices in the line string.
+   * @param zDepth Z value in local coordinates to use for each point.
+   */
+  public addLineString2d(points: Point2d[], zDepth: number): void {
+    const pts3d = copy2dTo3d(points, zDepth);
+    this.addLineString(pts3d);
+  }
+
+  /**
+   * Appends a 3d point string to the builder. The points are drawn disconnected, with a diameter in pixels defined by the builder's active [[GraphicParams.rasterWidth]].
+   * @param points Array of vertices in the point string.
+   */
+  public addPointString(points: Point3d[]): void {
+    this[_accumulator].addPointString(points, this.getLinearDisplayParams(), this.placement);
+  }
+
+  /**
+   * Appends a 2d point string to the builder. The points are drawn disconnected, with a diameter in pixels defined by the builder's active [[GraphicParams.rasterWidth]].
+   * @param points Array of vertices in the point string.
+   * @param zDepth Z value in local coordinates to use for each point.
+   */
+  public addPointString2d(points: Point2d[], zDepth: number): void {
+    const pts3d = copy2dTo3d(points, zDepth);
+    this.addPointString(pts3d);
+  }
+
+  /**
+   * Appends a closed 3d planar region to the builder.
+   * @param points Array of vertices of the shape.
+   */
+  public addShape(points: Point3d[]): void {
+    const loop = Loop.create(LineString3d.create(points));
+    this[_accumulator].addLoop(loop, this.getMeshDisplayParams(), this.placement, false);
+  }
+
+  /**
+   * Appends a closed 2d region to the builder.
+   * @param points Array of vertices of the shape.
+   * @param zDepth Z value in local coordinates to use for each point.
+   */
+  public addShape2d(points: Point2d[], zDepth: number): void {
+    const pts3d = copy2dTo3d(points, zDepth);
+    this.addShape(pts3d);
+  }
+
+  /**
+   * Appends a 3d open arc or closed ellipse to the builder.
+   * @param arc Description of the arc or ellipse.
+   * @param isEllipse If true, and if the arc defines a full sweep, then draw as a closed ellipse instead of an arc.
+   * @param filled If true, and isEllipse is also true, then draw ellipse filled.
+   */
+  public addArc(ellipse: Arc3d, isEllipse: boolean, filled: boolean): void {
+    let curve;
+    if (isEllipse || filled) {
+      curve = Loop.create(ellipse);
+    } else {
+      curve = Path.create(ellipse);
+    }
+
+    if (filled && !isEllipse && !ellipse.sweep.isFullCircle) {
+      const gapSegment: CurvePrimitive = LineSegment3d.create(ellipse.startPoint(), ellipse.endPoint());
+      (gapSegment as any).markerBits = 0x00010000; // Set the CURVE_PRIMITIVE_BIT_GapCurve marker bit
```

Comment:
is there any way around the cast to as any?

Reply:
It's been there since at least 2018, from the folks who initially ported this code from C++. It is the only occurrence of `markerBits` in the entire repository.

Reply:
Removed.